source/loader/layers/validation/ur_leak_check.hpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: MIT
+#ifndef UR_LEAK_CHECK_H
+#define UR_LEAK_CHECK_H 1
+
+#include "backtrace.hpp"
+#include "ur_validation_layer.hpp"
+
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+
+#define MAX_BACKTRACE_FRAMES 64
+
+namespace validation_layer {
+struct RefCountContext {
+  private:
+    struct RefRuntimeInfo {
+        int64_t refCount;
+        std::vector<BacktraceLine> backtrace;
+    };
+
+    enum RefCountUpdateType {
+        REFCOUNT_INCREASE,
+        REFCOUNT_DECREASE,
+    };
+
+    std::mutex mutex;
+    std::unordered_map<void *, struct RefRuntimeInfo> counts;
+
+    inline struct RefRuntimeInfo createRefInfo(int refCount) {
+        std::vector<BacktraceLine> backtrace = getCurrentBacktrace();
+
+        return RefRuntimeInfo{refCount, backtrace};
+    }
+
+    void updateRefCount(void *ptr, enum RefCountUpdateType type) {
+        std::unique_lock<std::mutex> ulock(mutex);
+
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            counts.emplace(ptr, createRefInfo(0));
+        }
+
+        switch (type) {
+        case REFCOUNT_INCREASE:
+            counts[ptr].refCount++;
+            break;
+        case REFCOUNT_DECREASE:
+            counts[ptr].refCount--;
+            break;
+        }
+
+        if (counts[ptr].refCount < 0 && type == REFCOUNT_DECREASE) {
+            context.logger.debug("Attempting to release nonexistent handle {}",
+                                 ptr);
+        }
+
+        context.logger.debug("Reference count for handle {} changed to {}", ptr,
```

Comment:
Please use different log levels for events depending on their expected occurrence.

Someone might want only to see the errors in their app, not every single refcount change.